[otbn] Document data integrity measures

imphil opened the pull request:
Start the documentation of security features in OTBN with a
documentation of its data integrity measures. This commit adds two new
sections to the documentation:

- A high-level security feature overview, i.e. the "sales pitch."
- A more detailed description of how the features are implemented in the
  Design Details section.

The documentation is a starting point and we will need to add more
features and refine them as we move on with specification and
implementation.

Fixes #4007

imphil commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+Both register files are able to accept protected data or unprotected data, and add integrity protection on demand before storing the data value.
+Wherever possible data with attached integrity data is consumed and emitted from the register file.
+
+In the following cases the register files consume integrity-protected data:
```
@GregAC can you please carefully read through this list? The bottom two bullet points are not implemented right now IIRC, and RND is missing completely (since we have not yet had much of a discussion there). This makes the document currently a bit of a mixed bag: some things are implemented, others are intended to be implemented, others are intended to be implemented but are missing currently. Maybe I can make that clearer by adding some "TODO: not yet implemented" parts to it?

rswarbrick replied:
I think that BN.SEL also copies the register bits across (judging by the notes in the RTL PR)?

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+It has a minimum Hamming distance of four, resulting in the ability to detect at least three errors in a 32 bit word.
+The code is used for error detection only, no error correction is performed.
+
+In contrast to the [original Hsiao code](https://arxiv.org/pdf/0803.1217.pdf) the outputs are inverted to produce a code in which the all-zero word (0x0) is not a valid code word.
```
Does this leave all ones as a valid code word? (And should it?)

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+
+#### Register File Integrity Protection
+
+OTBN contains two register files, the 32b GPRs, and the 256b WDRs.
```
Nit: I think this should probably be "two register files: the 32b GPRs and the 256b WDRs.". No need for the second comma either way, and the first makes it look like we've got 4 register files in total.

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+Both register files store error detection bits next to the actual data.
+
+The GPRs `x2` to `x31` store a 32b data word together with the integrity protection code, resulting in 39b of stored data.
+(The `x0`, the zero register, and `x1`, the call stack, require special treatment.)
```
Nit: Too many "the"s for x0!

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+#### Register File Integrity Protection
+
+OTBN contains two register files, the 32b GPRs, and the 256b WDRs.
+Both register files store error detection bits next to the actual data.
```
Should we explain why that's relevant here? (I actually don't know!)

imphil replied:
I didn't really intend to focus on the "next to the actual data" bit in this sentence, I reworded it.

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+Each 256b wide data register (WDR) stores a 256b data word together with the integrity protection code, resulting in 312b of stored data.
+The integrity protection is done separately for each of the eight 32b sub-words within a 256b word.
+
+Both register files are able to accept protected data or unprotected data, and add integrity protection on demand before storing the data value.
```
Hmm, I think that this needs slightly rejigging to make it clear that they only add check bits for unprotected data?

rswarbrick commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,43 @@ These accesses are ignored if OTBN is busy.
+The integrity protection is done separately for each of the eight 32b sub-words within a 256b word.
+
+Both register files are able to accept protected data or unprotected data, and add integrity protection on demand before storing the data value.
+Wherever possible data with attached integrity data is consumed and emitted from the register file.
```
Nit: Probably add a comma after "possible". Also, maybe it would be better to say "attached integrity bits" or something to avoid the double data? (Also, some people feel strongly about whether the word "data" is plural. I'm probably not one of them though :-))

GregAC commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -426,6 +450,42 @@ These accesses are ignored if OTBN is busy.
+The register files can consume integrity-protected data, or add protection to unprotected data.
+Whenever possible, integrity-protected data is consumed by the register files, in particular in the following cases:
+
+* Data coming from the data memory (DMEM) through the load-store unit.
```
```suggestion
* Data coming from the data memory (DMEM) through the load-store unit to a WDR or GPR.
```

As you said in the other comment (that github won't let me reply to) a TODO or similar here noting not all of these are implemented would be a good idea. I can think of at least one more:

* Data coming from the call stack to a GPR.

imphil replied:
I made these modifications and added a TODO to the ones that remain unimplemented after #6374 goes in.

GregAC replied:
Have just realised my bullet point is wrong! The issue being there's no instruction that directly transfers from call stack to/from a GPR, as it'll all go via the ALU (e.g. `addi x2, x1, 0` would be used to pop the call stack into `x2`).

We could special case things, e.g. so `addi xn, xy, 0` does preserve integrity.

I was also thinking if we had integrity on the PC with the same code we'd carry that directly to/from the call stack but of course how PC integrity is dealt with remains an open question.

imphil commented on hw/ip/otbn/doc/_index.md:
IMO yes, the detection probability increases by scrambling the integrity-protected data, since the scrambling diffuses error bits more. (It additionally makes it harder to pre-calculate which bits need to be flipped, but that's another property.)

However, my attempts to quickly come up with something that looks like a proper probability calculation failed, leaving us for the moment only with my gut feeling (and the impression that @msfschaffner did present this fact to me convincingly).

But we really should get out of the gut feeling domain here and do the math properly. @msfschaffner have you done that in the past and can share it, or should we have a look?

To avoid having to block this PR on that I'm removing the second part of the sentence for now, and we can add it back once we're sure about it.

rswarbrick replied:
Well, the point is that a small hamming weight change to an instruction can do exciting things (flipping an enable etc.), or similarly a data bool might change its meaning. But if you make a small hamming weight change to a scrambled value *and* happen to still hit a valid ECC codeword (which isn't trivial), you'll have smooshed stuff all over the value you're trying to change.

Another way to look at this is that if you have a specific change you want to make to the unscrambled data then, even if you know the scrambling key, it's extremely unlikely that the change you need to make to the scrambled data is small.

I'm not sure whether it's possible to to be more formal about this (with probabilities etc.) You'd need to have some measure over "interesting changes to data words", which seems like it's going to be very subjective.

imphil commented:
@tjaychen @cdgori I would appreciate if you could have a look at the specification proposed here. It matches what we have discussed before, but adds a bit more detail in some areas. It is not yet implemented (mostly) and also has a dependency on the system wide transmission integrity to ingest integrity-protected data through the bus window when storing data in OTBN's instruction and data memories. We can discuss some of that tomorrow in the meeting. Thanks!

cdgori commented:
> @tjaychen @cdgori I would appreciate if you could have a look at the specification proposed here. It matches what we have discussed before, but adds a bit more detail in some areas. It is not yet implemented (mostly) and also has a dependency on the system wide transmission integrity to ingest integrity-protected data through the bus window when storing data in OTBN's instruction and data memories. We can discuss some of that tomorrow in the meeting. Thanks!

I had a look at this and it generally seems to match what we discussed before.  

I had one question about the wording for the section discussing the preservation of integrity protection, especially for `BN.MOV` / `BN.MOVR` instructions.  I would have assumed that copying a wide data from one address to another would have triggered the bijective address mapping of the scrambling and therefore resulted in a different value actually being written to the register file, in order to hinder aliasing/remapping attacks (which are also protected by integrity checks on the contents of the instruction memory as well).

So I am assuming that the text is specifically referring to just the preservation of the Hsiao code, not the actual value that is ultimately stored in the regfile (including any scrambling)?  I originally read the text, possibly in a caffeine-reduced state :), and got a bit concerned that at a high-level it's possible to have some kind of aliasing attack if the same value appeared in two regfile locations.  But I believe I convinced myself that this section is only talking about the integrity bit preservation, not the integrity + scramble - it would be good to confirm that though.  In general the idea of preserving the integrity end-to-end is desirable, so it's good to see this principle being applied.

I guess semi-related: I also don't see an explicit statement of the order of the transformations applied to the data, which might be good to have here - my assumption is that it's (integrity gen/preserve -> scramble) for writes and (descramble -> integrity check) for reads?  It might be implied from an OTBN block diagram somewhere, but I can't find one to look at, at this very instant.

If I've misunderstood the order and the integrity generation is after the scramble for writes, then I'll have to think about this again.

cdgori reviewed (changes requested):
Need to add a sentence or two clarifying the order of scramble and integrity generate (and also descramble / integrity check) to this document to make it clear - see longer comment above.

msfschaffner commented:
@imphil RE your question regarding scrambling of integrity protected data:

Thus scrambling really has two goals: 
1. obfuscation of memory contents to make passive probing attacks more difficult, 
2. diffuse data to make active fault injection attacks more difficult.

Number 2. above can be either directed, or completely random. I.e., an attacker can try to inject specific values, or he can try glitch data at random.

As for injecting specific values, I think @rswarbrick explained quite well why it makes sense to scramble both ECC+data. The additional diffusion layer of the scrambling makes it harder for an attacker to inject consistently encoded words by just controlling a few bits / injecting random glitches at the memory boundary.
So in order to mount a successful attack, an attacker would need access to all (or a large subset) of the data bits, AND he would have to know the scrambling key.

As for injecting random values, I think the scrambling does not significantly alter the ECC detection probability. I.e., if you assume the perturbed words follow a uniform distribution, then the data after passing through the scrambler should also follow a uniform distribution - and therefore the ECC detection probability should remain the same.

So I think for randomly injected faults, the motivation is more along the lines of what @GregAC said: the diffusion ensures that even small hamming weight faults results in large changes after the scrambler, and the non-linear transformation with an ephemeral key makes it harder to predict the corrupted values after the scrambler.

(@cdgori please correct me if you think some of my reasoning is flawed).

cdgori commented:
That all looks correct to my understanding @msfschaffner - and thanks to @rswarbrick and @GregAC for the earlier comments that you referred to.  I believe it is ok to assume that the perturbed words follow a uniform distribution - but I would be open to hearing arguments about potential bias/non-uniformity in the distribution too.  It's possible that 0->1 (or 1->0) is easier to implement (or more likely to occur) for certain types of physical structures, I just don't know enough here to say if that is the case.

Note that the increased difficulty in case 1 (obfuscation) is pretty small in reality for an attacker with strong passive probing capability (it only stops a certain class of adversary that has the ability to probe the wide memories but somehow not the key register as well, which is a pretty unlikely adversary), but we get it basically "for free" because we have to deal with case 2, so we will gladly take the small improvement.

(That same argument roughly applies to case 2 for specific glitch values, except I think there are real physical limits on the number of active probes that can be placed due to mechanical issues of landing bond wires/probe pads in a relatively small area.  Sampling the key value can probably be done passively by the adversary, it just complicates the attack since it requires both passive sampling and active injection to be successful.)

imphil commented:
Thanks for your comments @cdgori and @msfschaffner. I updated the text to hopefully clarify some questions that came up.

> I also don't see an explicit statement of the order of the transformations applied to the data, which might be good to have here

There was a bit of language explaining that in the data/instruction memory integrity protection sections, but when I re-read it, it wasn't really clear enough. I adjusted those paragraphs to be more explicit. When storing data, the ECC is calculated first, and then the data+ECC word is scrambled.

> I am assuming that the text [about preserving integrity protection] is specifically referring to just the preservation of the Hsiao code

Yes, that's correct. I made the text more explicit (hopefully).

Further changes:

* I now also mention address scrambling for imem/dmem addresses in the text explicitly, something that was missing before.
* I removed the mention of an the *inverted* Hsiao code. This is something we talked about and might potentially do, but nothing we implement at the moment or even have fully agreed on implementing. Let's reduce the confusion for the reader of the docs by making it factually correct as of today. Further discussion is tracked in https://github.com/lowRISC/opentitan/issues/6478.
* A couple of editorial cleanups.

@cdgori would you mind taking another look?

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -499,6 +523,107 @@ This signal is high when OTBN is not running.
+### Data Integrity Protection {#design-details-data-integrity-protection}
+
+OTBN stores and operates on data (state) in its dedicated memories, register files, and internal registers.
+OTBN's data integrity protection is designed to protected all data stored and transmitted within OTBN from modifications through physical attacks.
```
protected -> protect

cdgori commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -499,6 +523,107 @@ This signal is high when OTBN is not running.
+#### Memory Scrambling {#design-details-memory-scrambling}
+
+Contents of OTBN's instruction and data memories are scrambled while at rest.
+Both the data word itself and the address are scrambled.
```
I can interpret this sentence two ways:

1) data is scrambled, and (separately) address is scrambled
2) data and address are together inputs to the scramble function (i.e. address affects the data written / ~binds the data to the specific address, and also affects the actual storage address in the memory)

I'm not sure which is the correct meaning here (I think it's 2?)

GregAC replied:
It's 2. The keystream that gets XORed with the data is generated with an initial vector derived from the address and nonce. The address itself goes through the substitution/permutation network to form the physical memory address.

imphil replied:
I updated the text to give a bit more detail here without duplicating all the gory details, which are in the SRAM controller spec linked further down.

cdgori reviewed (changes requested):
This is looking really good now @imphil - one nit and one sentence I would like just slightly clarified and I think you are ready to go.

tjaychen reviewed (approved):
this is great!

GregAC commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -499,6 +523,107 @@ This signal is high when OTBN is not running.
+Whenever possible the Integrity Protection Code is preserved from its source and written directly to the register files without recalculation, in particular in the following cases:
+
+* Data coming from the data memory (DMEM) through the load-store unit to a GPR or WDR.
+  (TODO: Not yet implemented.)
```
This has been implemented

GregAC commented on hw/ip/otbn/doc/_index.md:
```diff
@@ -499,6 +523,107 @@ This signal is high when OTBN is not running.
+OTBN's data integrity protection is designed to protected all data stored and transmitted within OTBN from modifications through physical attacks.
+
+During transmission, the integrity of data is protected with an integrity protection code.
+Data at rest in the instruction and data memories is additionally scrambled.
```
Is it worth clarifying here or elsewhere that it's the full data + integrity that gets scrambled and stored in memory?

imphil replied:
There is a more detailed description of that in the sections "Data Memory (DMEM) Integrity Protection" and "Instruction Memory (IMEM) Integrity Protection" below. This initial paragraph is only meant as a high-level intro. Let me know if the description further down is unclear or can be extended.

imphil commented:
Thanks for your comments! I updated the PR to hopefully reflect those. It's a bit tricky to keep some of the sections high-level enough to serve as intro without confusing readers at the same time (i.e. the eternal question of "how far can we abstract").

cdgori reviewed (approved):
Thanks @imphil and @GregAC - LGTM!

imphil commented:
Thanks a lot @cdgori for your helpful reviews, they made the document much better.

cdgori commented:
> Thanks a lot @cdgori for your helpful reviews, they made the document much better.

No problem, happy I could help.  I fully agree with your last/previous comment on this PR about how tricky it is to keep the writing high-level/intro but not lose the reader by abstracting too much.  There's also a matter of personal preference / stylistic liberty too.

Documentation is hard, and I feel like OT has done a good job producing high-quality docs.  We should all be quite happy about that.
